scalaparse/shared/src/main/scala/scalaparse/Xml.scala

```diff
@@ -11,120 +11,108 @@ trait Xml extends Core {
-
-    val AttValue = P(
-      "\"" ~ (CharQ | Reference).rep ~ "\"" |
-      "'" ~ (CharA | Reference).rep ~ "'" |
-      ScalaExpr
-    )
-
-    val Comment = P( "<!--" ~ ((!"-" ~ Char) | ("-" ~ (!"-" ~ Char))).rep ~ "-->" )
+    val Comment = P( "<!--" ~/ ((!"-" ~ Char) | ("-" ~ (!"-" ~ Char))).rep ~ "-->" )
 
-    val PI = P( "<?" ~ PITarget ~ (WL ~ (!"?>" ~ Char).rep).? ~ "?>" )
+    val PI = P( "<?" ~/ PITarget ~ (WL ~ (!"?>" ~ Char).rep).? ~ "?>" )
+    val PIProcText = P( WL ~ (!"?>" ~ Char).rep )
     val PITarget = P( !(("X" | "x") ~ ("M" | "m") ~ ("L" | "l")) ~ Name )
-    val CharRef = P( "&#" ~ CharIn('0' to '9').rep(1) ~ ";" | "&#x" ~ Basic.HexNum ~ ";" )
+
     val Reference = P( EntityRef | CharRef )
-    val EntityRef = P( "&" ~ Name ~ ";" )
-    val ScalaExpr = P("{" ~ WS ~ Block ~ WL ~ "}")
-    val Char = P( AnyChar )
-    val CharData = P( (!("{" | "]]>" | CharRef) ~ Char1 | "{{").rep(1) )
+    val EntityRef = P( "&" ~ Name ~/ ";" )
+    val CharRef = P( "&#" ~ Num ~/ ";" | "&#x" ~ HexNum ~/ ";" )
+    val Num = P( CharIn('0' to '9').rep )
+    val HexNum = P( CharIn('0' to '9', 'a' to 'f', 'A' to 'F').rep )
 
+    val CharData = P( (!"{" ~ Char1 | "{{").rep(1) )
+
+    val Char = P( AnyChar )
     val Char1  = P( !("<" | "&") ~ Char )
     val CharQ = P( !"\"" ~ Char1 )
     val CharA = P( !"'" ~ Char1 )
-    val CharB = P( !"{" ~ Char1 )
-    val Name = P( XNameStart ~ NameChar.rep )
-    val XNameStart  = P( "_" | BaseChar | Ideographic )
-
-    val NameStartChar = P(CharIn(
-      ":", 'A' to 'Z', "_", 'a' to 'z', '\u00C0' to '\u00D6', '\u00D8' to '\u00F6',
-      '\u00F8' to '\u02FF', '\u0370' to '\u037D', '\u037F' to '\u1FFF', '\u200C' to '\u200D',
-      '\u2070' to '\u218F', '\u2C00' to '\u2FEF', '\u3001' to '\uD7FF', '\uF900' to '\uFDCF',
-      '\uFDF0' to '\uFFFD' // ++ [#x10000-#xEFFFF] ???? don't chars max out at \uffff ????
-    ))
-
-    val NameChar = P( NameStartChar | CharIn(
-      "-", ".", '0' to '9', "\u00B7", '\u0300' to '\u036F', '\u203F' to '\u2040'
-    ))
-    val ElemPattern: P0 = P( TagPHeader ~ (EmptyElemTagPEnd | STagPEnd ~ ContentP ~ ETagP ))
-    val TagPHeader = P( "<" ~ Name ~ WL.?  )
-    val EmptyElemTagPEnd = P( "/>" )
-    val STagPEnd = P( ">")
-    val ETagP = P( "</" ~ Name ~ WL.? ~ ">" )
-    val ContentP = P( CharData.? ~ ((ElemPattern | ScalaPatterns) ~ CharData.?).rep )
-    val ContentP1 = P( ElemPattern | Reference | CDSect | PI | Comment | ScalaPatterns )
+
+    val Name = P( NameStart ~ NameChar.rep )
+    val NameStart  = P( CharPred(isNameStart) )
+    val NameChar = P( CharPred(isNameChar) )
+
+    val ElemPattern: P0 = P( TagPHeader ~/ ("/>" | ">" ~/ ContentP ~/ ETag ) )
+    val TagPHeader = P( "<" ~ TagName ~ WL.?  )
+
+    val ContentP: P0 = P( (ElemPattern | CharDataP | ScalaPatterns).rep )
     val ScalaPatterns = P( "{" ~ Patterns ~ WL ~ "}" )
+    val CharDataP = P( "&" ~ CharData.? | CharData ) // matches weirdness of scalac parser on xml reference.
```

Comment:
I think this is a bug in scalac. A better solution might be to fix the scalac parser instead